### DISCUSSION TITLE
Create door with lining in python

### DISCUSSION THREAD

**Discussion by NkStll** [IfcOpenShell/IfcOpenShell#2863] [Q&A]
Hi,
I'm struggling to generate an ifc file, that contains a door with a lining in python using ifcopenshell.
I am able to generate a door by

```
door = run("root.create_entity", model, ifc_class="IfcDoor",name="door")
representation = run("geometry.add_wall_representation", model, context=body, length=0.95, height=2.1, thickness=0.05)
run("geometry.assign_representation", model,product=door, representation=representation)
run("style.assign_representation_styles", model, shape_representation=representation)
run("geometry.edit_object_placement", model, product=door,element=opening, matrix=matrix_io)
run("void.add_filling", model, opening=opening, element=door)
run("spatial.assign_container", model, product=door, relating_structure=storey2)
```

But I do not understand how to add a door frame with the help of _IfcDoorLiningProperties_ and I can't seem to find any coding examples.
Thank you for your help!

**Comment by aothms** [comment]
I don't think there is high level API functionality for that (`run()`). As a rough sketch you could try the following:

```python
model.createIfcRelDefinesByProperties(
    ifcopenshell.guid.new(),
    model.by_type('IfcOwnerHistory')[0],
    RelatedObjects=[door],
    RelatingPropertyDefinition=model.createIfcDoorLiningProperties(...)
)
```

**Reply by Moult** [reply]
Actually surprise there actually is a high level API function for that. Just that it isn't wired up to listen to the predefined properties like IFC door lining properties (yet?). Here it is: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/geometry/add_door_representation.py

You can see example parameters collected here https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/module/model/door.py#L47-L73
